ceno_zkvm/src/instructions/global.rs

```diff
@@ -0,0 +1,817 @@
+    perm: P,
+}
+
+impl<E: ExtensionField, P> GlobalConfig<E, P> {
+    // TODO: make `WIDTH`, `HALF_FULL_ROUNDS`, `PARTIAL_ROUNDS` generic parameters
+    pub fn configure(
+        cb: &mut CircuitBuilder<E>,
+        rc: RoundConstants<E::BaseField, 16, 4, 13>,
+        perm: P,
+    ) -> Result<Self, CircuitBuilderError> {
+        let x: Vec<WitIn> = (0..SEPTIC_EXTENSION_DEGREE)
+            .map(|i| cb.create_witin(|| format!("x{}", i)))
+            .collect();
+        let y: Vec<WitIn> = (0..SEPTIC_EXTENSION_DEGREE)
+            .map(|i| cb.create_witin(|| format!("y{}", i)))
+            .collect();
+        let slope: Vec<WitIn> = (0..SEPTIC_EXTENSION_DEGREE)
+            .map(|i| cb.create_witin(|| format!("slope{}", i)))
+            .collect();
+        let addr = cb.create_witin(|| "addr");
+        let is_ram_register = cb.create_witin(|| "is_ram_register");
+        let value = UInt::new(|| "value", cb)?;
+        let shard = cb.create_witin(|| "shard");
+        let global_clk = cb.create_witin(|| "global_clk");
+        let local_clk = cb.create_witin(|| "local_clk");
+        let nonce = cb.create_witin(|| "nonce");
+        let is_global_write = cb.create_witin(|| "is_global_write");
+
+        let is_ram_reg: Expression<E> = is_ram_register.expr();
+        let reg: Expression<E> = RAMType::Register.into();
+        let mem: Expression<E> = RAMType::Memory.into();
+        let ram_type: Expression<E> = is_ram_reg.clone() * reg + (1 - is_ram_reg) * mem;
+        let perm_config = Poseidon2Config::construct(cb, rc);
+
+        let mut input = vec![];
+        input.push(addr.expr());
+        input.push(ram_type.clone());
+        // memory expr has same number of limbs as register expr
+        input.extend(value.memory_expr());
+        input.push(shard.expr());
+        input.push(global_clk.expr());
+        // add nonce to ensure poseidon2(input) always map to a valid ec point
+        input.push(nonce.expr());
+        input.extend(repeat_n(E::BaseField::ZERO.expr(), 16 - input.len()));
+
+        let mut record = vec![];
+        record.push(addr.expr());
+        record.push(ram_type);
+        record.extend(value.memory_expr());
+        record.push(local_clk.expr());
+
+        // if is_global_write = 1, then it means we are propagating a local write to global
+        // so we need to insert a local read record to cancel out this local write
+        cb.assert_bit(|| "is_global_write must be boolean", is_global_write.expr())?;
+        // TODO: for all local reads, enforce they come to global writes
+        // TODO: for all local writes, enforce they come from global reads
+
+        // global read => insert a local write with local_clk = 0
+        cb.condition_require_zero(
+            || "is_global_read => local_clk = 0",
+            1 - is_global_write.expr(),
+            local_clk.expr(),
+        )?;
+        // TODO: enforce shard = shard_id in the public values
```

Comment:
there is api `cb.query_shard_id()` and we need below constrian
- if is_global_write == 0, then assert_lt(global_clk, cb.query_shard_id())
- if is_global_write == 1, then assert_eq(global_clk, cb.query_shard_id())